lib/pbench/agent/tool_data_sink.py

```diff
@@ -738,6 +739,21 @@ class ToolDataSinkParams(NamedTuple):
     tool_trigger: str
     tools: Dict[str, str]
 
+    def __str__(self) -> str:
+        """A string containing a deterministic representation of the params
+
+        Convert the field names to dict keys, produce them in sorted order, and
+        filter out any object references, because their addresses vary from run
+        to run.
+        """
+        return str(
+            {
+                k: v
+                for k, v in sorted(self._asdict().items())
+                if isinstance(v, (dict, list, tuple, str, int, float, bool, type(None)))
+            }
+        )
```

Comment:
In Java there's a standard utility class that builds a generic string representation of an object by introspection for precisely this purpose (it's inefficient, but often convenient); I wonder if Python has such a thing? (Note, a quick search didn't reveal such a beast, unless it's nested within `inspect`.) If not, perhaps you could factor your two loops out into our own "generic string" helper?

The `NamedTuple` "class" is actually a weird meta-class, and I'm not sure you could get away with building a common base class for these two. (I remember looking at one point, and documentation suggested that was a "no".) At least, I think it would require more than the normal annotation.

Reply:
Python is happy to produce a string representation, via either `__str__()` or `__repr__()`, depending on "details"...but they tend to end up with addresses in them, which is what I'm trying to avoid.  (I _could_ take their output and apply a regex substitution to them, but that seems...heavy-handed?)

I'll see how my next attempt looks.

Reply:
OK, what I've got is, I believe, adequate for the immediate need.  It's nicely encapsulated, so we can expand upon or replace it easily in the future.